Feature - Add additional/dynamic calls as cli arguments

paulwuertz opened the pull request:
Enables the user to add dynamic calls like callbacks to the call graph.

Closes https://github.com/HBehrens/puncover/issues/105.

Add the funcion call from func_a calling func_1 by `--add-dynamic-calls 'func_a->func_1'`.

An example the cannectivity firmware based on zephyr.

It has a driver abstraction calling the chip-specific driver layer, which then calls a user given callback function. In the images below user_callback and driver_implementation have both no callers(1), as they are inserted as function pointers to some structs at runtime. The os_interface function has 3 callees. The worst case max stack recognized for the gs_usb_tx_thread is 712 bytes where the call tree goes up to the os_interface `z_impl_can_send` and the callback worst case stack usage is 488 bytes(2).

|              user_callback          |     driver_implementation | os_interface   |
|:-------------------------------------:|:-------------------------------------:|:-------------------------------------:|
| ![user_callback_without_dynamic_calls](https://github.com/user-attachments/assets/5c73b1c6-04fc-4095-9ff1-8ccc1137392c) | ![can_mcan_send_without_dynamic_calls](https://github.com/user-attachments/assets/1e55b845-2e9a-4204-922c-0b70e2284df4) | ![z_impl_can_send_without_dynamic_calls](https://github.com/user-attachments/assets/d08b6453-473f-49d4-9503-54d349e6419c) |

When adding the 3 dynamic calls os_interface->driver_implementation->user_callback as arguments:

* --add-dynamic-calls "z_impl_can_send->can_mcan_send"
* --add-dynamic-calls "can_mcan_send->gs_usb_can_tx_callback"

```bash
puncover --elf_file ~/git/cannectivity/build/zephyr/zephyr.elf --build_dir ~/git/cannectivity/build/  --add-dynamic-calls "z_impl_can_send->can_mcan_send" --add-dynamic-calls "can_mcan_send->gs_usb_can_tx_callback"
```

Then we get user_callback and driver_implementation having one caller (3) getting called from z_impl_can_send (4) and all are in the same worst stack path with now using up to 856 bytes.


|              user_callback          |     driver_implementation | os_interface   |
|:-------------------------------------:|:-------------------------------------:|:-------------------------------------:|
| ![user_callback_with_dynamic_calls](https://github.com/user-attachments/assets/ad52f9ab-5f33-452d-8b99-cb6d9caa1a8a) | ![Screen Shot 2025-02-16 at 19 49 32](https://github.com/user-attachments/assets/60365c0f-1e09-4c3f-a4a7-02a3711c6c2b) | ![Screen Shot 2025-02-16 at 19 49 19](https://github.com/user-attachments/assets/c92b7085-95d1-4e11-8aa6-40d8b4af2f55) |

With the open PR https://github.com/HBehrens/puncover/pull/106 for the conig file it looks like this:

```yaml
elf_file: ~/git/cannectivity/build/zephyr/zephyr.elf   
build_dir: ~/git/cannectivity/build/
add-dynamic-calls: [
    z_impl_can_send->can_mcan_send,
    can_mcan_send->gs_usb_can_tx_callback,
]
```

paulwuertz committed:
add dynamic calls as cli arguments, closes #105

paulwuertz committed:
handle empty list of manually added dynamic calls

noahp commented:
@paulwuertz let me know how you'd like to proceed with this one! I think it more or less makes sense to me (interesting use case indeed), needs some rebasing and tests ideally

paulwuertz committed:
add test for dynamic_call fn

paulwuertz committed:
add exit+tests exiting with msg+errorcode when a dynamic call to unknown function of unknown name is added

paulwuertz commented:
> @paulwuertz let me know how you'd like to proceed with this one! I think it more or less makes sense to me (interesting use case indeed), needs some rebasing and tests ideally

Rebased and added a simple test happy path and error test case. Seems ok or any other ideas :)?

I was unsure, weather by symbol name would be enough, or filepath was necessary as well.
Maybe there could be name conflicts with module scoped static functions. I guess this could be an edge case.
For most firmware I suppose this is not a thing and for global C and namespaced C++ functions it should be ok as a first shot?

paulwuertz commented:
I also sneaked in a commit calling the export function for functions+variables to the json report.
It was removed and was an unused function now - I could also open another PR for that commit if prefered :)

-> moved it to https://github.com/HBehrens/puncover/pull/147 :)

paulwuertz committed:
Merge branch 'master' into add_dynamic_calls_via_cli

paulwuertz committed:
fix cyclic symbol endless loop on error print

paulwuertz commented:
Some test results playing around with zephyr with the console and compared to the thread analyzer:

> puncover --build_dir ~/zephyrproject/zephyr/build/ --non-interactive --generate-report --report-tag console --report-filename console --report-max-static-stack-usage  shell_thread --elf ~/zephyrproject/zephyr/build/zephyr/zephyr.elf

reported max_static_stack_size: 232 - but on the thread analyzer we are already after boot at 368 bytes...

After pressing enter - 600 bytes used, after entering `help` 976 bytes used...

Now when I add just two dynamic calls...

> puncover --build_dir ~/zephyrproject/zephyr/build/ --generate-report --report-tag console --report-filename console --report-max-static-stack-usage  shell_thread **--add-dynamic-calls "shell_signal_handle->shell_process" --add-dynamic-calls "shell_signal_handle->kill_handler"** --elf ~/zephyrproject/zephyr/build/zephyr/zephyr.elf 

... then we are much closer with 832 to the worst case found on the target - being found along the dynamic call given. 

But it underreports still by ~140 bytes. 

Well, there is a lot of device callbacks in zephyr, so a nice use-case for this feature I think :)

{
  

>   "console": {
>         "timestamp": "2026-05-09T00:10:13.747823",
>         "stack_report": {
>             "shell_thread": {
>                 "max_static_stack_size": 832,
>                 "call_stack": [
>                     {
>                         "function": "shell_thread",
>                         "name": "shell_thread",
>                         "stack_size": 32
>                     },
>                     {
>                         "function": "shell_signal_handle",
>                         "name": "shell_signal_handle",
>                         "stack_size": 24
>                     },
>                     {
>                         "function": "shell_process",
>                         "name": "shell_process",
>                         "stack_size": 8
>                     },
>                     {
>                         "function": "state_collect",
> ...

paulwuertz commented:
Closed in favor of https://github.com/HBehrens/puncover/pull/158
